### ISSUE TITLE
Interoperability Bug when using Elliptic Curve CA

### ISSUE THREAD

**Issue by eccgecko** [ARMmbed/mbedtls#1411]
### Description
- Type: Bug
- Priority: Major

---------------------------------------------------------------
## Bug

TLS Handshake fail when using Elliptic Curve certs/keys created using a CA signed with RSA.

**OS**  
Linux Ubuntu | iOS

**mbed TLS build:**  
Version: 2.5.1  
OS version: Ubuntu 17.10 | iOS 11.2.6  

Compiler and options (if you used a pre-built binary, please indicate how you obtained it): 

Ubuntu: apt-get install libmbedtls-dev

iOS: library built into OpenVPN Connect app

**Expected behavior**   
When using OpenVPN built with mbedTLS library, it is expected to be able to connect to a server using ECDSA and ECDHE with certs/keys built using EasyRSA.

**Actual behavior**  
EasyRSA is creating EC CAs using 2048 RSA as the signing algorithm. I am not sure if this is a bug or not (I have reported it [here](https://github.com/OpenVPN/easy-rsa/issues/196) ).

However, in theory this should still not interfere with the TLS handshake that occurs when the VPN is establishing a connection. This is shown to be the case because, when using OpenVPN with OpenSSL 1.0.2n, the TLS handshake is successful and is using ECDSA and ECDHE:
`Control Channel: TLSv1.2, cipher TLSv1/SSLv3 ECDHE-ECDSA-AES256-GCM-SHA384, 384 bit EC, curve: secp384r1`

When using mbedTLS, the handshake fails.
On iOS, the log shows
`2018-02-25 15:12:22 VERIFY FAIL -- The certificate is signed with an unacceptable key (eg bad curve, RSA too short). : depth=0`
`2018-02-25 15:12:22 Transport Error: mbed TLS: SSL read error : X509 - Certificate verification failed, e.g. CRL, CA or signature check failed
2018-02-25 15:12:22 EVENT: CERT_VERIFY_FAIL mbed TLS: SSL read error : X509 - Certificate verification failed, e.g. CRL, CA or signature check failed [ERR]`

On Ubuntu, the log shows
`Mon Feb 26 13:02:44 2018 VERIFY ERROR: depth=0, subject=CN=iOSServer2: The certificate is signed with an unacceptable key (eg bad curve, RSA too short).
Mon Feb 26 13:02:44 2018 TLS_ERROR: read tls_read_plaintext error: X509 - Certificate verification failed, e.g. CRL, CA or signature check failed
Mon Feb 26 13:02:44 2018 TLS Error: TLS object -> incoming plaintext read error
Mon Feb 26 13:02:44 2018 TLS Error: TLS handshake failed
Mon Feb 26 13:02:44 2018 SIGUSR1[soft,tls-error] received, process restarting`

**Steps to reproduce**  

Create CA and Cert/Key pairs using latest EasyRSA 3.0.5 and selecting EC as the algorithm. 
Use the created CA and Client/Key pairs to create VPN server and client config, then use mbedTLS to connect to server.

**Comment by ordex** [comment]
This is probably an ill-formed PKI, as I don't think anybody expects to have the CA using RSA and the certs using EC, however, it is interesting to note that OpenSSL is able to accept this while mbedTLS does not.

**Comment by sbutcher-arm** [comment]
Thanks for the feedback! Offhand I'm not sure what's going on here, but we'll certainly look into it.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2158

**Comment by hanno-arm** [comment]
Hi @eccgecko,

first, let me apologise that we haven't been responsive on this issue until now.

> EasyRSA is creating EC CAs using 2048 RSA as the signing algorithm. I am not sure if this is a bug or not.

Who's the signer for the issued CRT? Is it a self-signed certificate? If not, there's no reason why the algorithm used to sign the CRT should match type of public key within the CRT, and Mbed TLS does not require that. 

Did you disable RSA in your configuration of Mbed TLS?

Kind regards,
Hanno

**Comment by hanno-arm** [comment]
Hi @eccgecko,

could you provide more information on the software that's running on client and server and which certificates are configured on both sides and how precisely they were generated? Is it correct that OpenVPN is running on both client and server? If so, which CRT profile is configured? (See below)

As far as I understand https://github.com/OpenVPN/easy-rsa/issues/196, you generated a CA with 2048-bit RSA-key and used that to sign a child CRT with an ECC key using `secp384r1`. While not what you  intended, I agree with 
> However, in theory this should still not interfere with the TLS handshake that occurs when the VPN is establishing a connection.

in that this should result in a valid CRT chain usable for the client. 

To check, I used the parent of https://github.com/OpenVPN/easy-rsa/commit/02532b231182d0d7c5794c526d2579418cccae1b (https://github.com/OpenVPN/easy-rsa/commit/02532b231182d0d7c5794c526d2579418cccae1b itself fixes the issue that EasyRSA is using RSA despite being configured for ECC) to generate an RSA-CA and two ECC-childs for client and server, respectively, and used those CRT to successfully setup a mutually authenticated session running Mbed TLS on both client and server. So, I cannot reproduce your issue at the moment.

Your log shows Mbed TLS' CRT verification fails on the end-entity CRT with failure flag `MBEDTLS_X509_BADCERT_BAD_KEY`. This is only set if 
* An RSA key is used but its bitlength is too low.
* An ECC key is used by its underlying curve is not allowed.

What is allowed or not is determined by the X.509 profile used by Mbed TLS, and which is accessible from the OpenVPN command line through the parameter `-tls-cert-profile`. For the default value `legacy` as well as `preferred`, 2048-bit RSA keys and secp384r1 ECC keys are allowed in Mbed TLS. However, for `-tls-cert-profile suiteb` will not allow RSA.

Which configuration of `-tls-cert-profile` are you using?

Kind regards,
Hanno

**Comment by eccgecko** [comment]
Hi @hanno-arm 

Apologies, but this was first reported back in February, and I already went and made my own work arounds, so cannot conduct further tests to help you with this, as the configuration I had is no longer on my system, and I don't really remember what the issue was now, or its solution.

What I do remember is that I was using self-signed certs made by Easy-RSA. An older version of Easy-RSA was using the RSA signing algorithm, even on EC certs.

This was causing an issue when trying to use these certs in openvpn when built using the mbedtls library.

It was not occurring when using openvpn built against openssl. 

When openvpn server was built against openssl, the certs loaded fine, but clients with mbedtls would fail. These were standard configs including the openvpn ios app. Clients with openssl did not fail, and could connect successfully. When I then built the server against mbedtls, the openvpn server would fail to load the certs.

I worked around this by ensuring the self-signed certs were signed using ecdsa, which stopped the bug with mbedtls.

Sorry I cant be of more use.
